# LIMPIEZA DE DATOS:

In [ ]:
#VERIFICAR TIPO DE CODIFICACION, en caso no sea UTF-8
import chardet

with open('Produccion_Agricola_Region_Cusco_2018-2020.csv', 'rb') as f:
    result = chardet.detect(f.read(100000))  # Lee una parte del archivo
print(result['encoding'])

ISO-8859-1


In [ ]:
#IMPORTAMOS LOS DATOS
import pandas as pd
#importar el csv
df_produccion = pd.read_csv('Produccion_Agricola_Region_Cusco_2018-2020.csv', sep= ';', encoding='ISO-8859-1')
print(df_produccion.head())

  DEPARTAMENTO PROVINCIA   DISTRITO  UBIGEO  PERIODO_AGRICOLA  \
0        CUSCO      ANTA  ANCAHUASI   80302              2020   
1        CUSCO      ANTA  ANCAHUASI   80302              2020   
2        CUSCO      ANTA  ANCAHUASI   80302              2020   
3        CUSCO      ANTA  ANCAHUASI   80302              2020   
4        CUSCO      ANTA  ANCAHUASI   80302              2020   

              CULTIVO  SUPERFICIE_VERDE  SIEMBRA  COSECHA  RENDIMIENTO  \
0             ALFALFA        190.470588    147.0    147.0    81972.789   
1   ARVEJA GRANO SECO         20.700000     23.0     23.0     2000.000   
2  ARVEJA GRANO VERDE         29.636364     40.0     31.0     6000.000   
3         AVENA GRANO         85.222222     98.0     98.0     2000.000   
4   CEBADA  FORRAJERA          8.117647     35.0     32.0    19875.000   

   PRODUCCION  PRECIO_CHACRA  SUPERFICIE_PERDIDA  VALOR_PRODUCCION  \
0     12050.0          0.190                 NaN          2289.500   
1        46.0          2

In [ ]:
#información del dataset
print(df_produccion.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4869 entries, 0 to 4868
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   DEPARTAMENTO        4869 non-null   object 
 1   PROVINCIA           4869 non-null   object 
 2   DISTRITO            4869 non-null   object 
 3   UBIGEO              4869 non-null   int64  
 4   PERIODO_AGRICOLA    4869 non-null   int64  
 5   CULTIVO             4869 non-null   object 
 6   SUPERFICIE_VERDE    4855 non-null   float64
 7   SIEMBRA             3975 non-null   float64
 8   COSECHA             4558 non-null   float64
 9   RENDIMIENTO         4555 non-null   float64
 10  PRODUCCION          4698 non-null   float64
 11  PRECIO_CHACRA       4698 non-null   float64
 12  SUPERFICIE_PERDIDA  93 non-null     float64
 13  VALOR_PRODUCCION    4869 non-null   float64
 14  FECHA_CORTE         4869 non-null   int64  
dtypes: float64(8), int64(3), object(4)
memory usage: 570.7+

In [ ]:
# Contar columnas completamente vacías
columnas_vacias = df_produccion.isnull().all()  # Devuelve True para columnas completamente vacías
cantidad_columnas_vacias = columnas_vacias.sum()

print(f"Número de columnas completamente vacías: {cantidad_columnas_vacias}")

Número de columnas completamente vacías: 0


In [ ]:
print(df_produccion.isnull().sum()) #muestra la cantidad de valores nulos por columna


DEPARTAMENTO             0
PROVINCIA                0
DISTRITO                 0
UBIGEO                   0
PERIODO_AGRICOLA         0
CULTIVO                  0
SUPERFICIE_VERDE        14
SIEMBRA                894
COSECHA                311
RENDIMIENTO            314
PRODUCCION             171
PRECIO_CHACRA          171
SUPERFICIE_PERDIDA    4776
VALOR_PRODUCCION         0
FECHA_CORTE              0
dtype: int64


In [ ]:
print(f"Cantidad de valores únicos en CULTIVO:", df_produccion['CULTIVO'].nunique())
print(df_produccion['CULTIVO'].value_counts())

Cantidad de valores únicos en CULTIVO: 73
CULTIVO
PAPA (agrupa mejoradas y nativas)    277
MAIZ AMILACEO                        255
QUINUA                               247
HABA GRANO SECO                      233
ALFALFA                              231
                                    ... 
TE                                     3
NUNA GRANO SECO                        2
SANDIA                                 2
LINAZA                                 1
MARANON                                1
Name: count, Length: 73, dtype: int64


 **Análisis de valores nulos**

**Columnas con pocos valores nulos:**

**SUPERFICIE_VERDE** (14 nulos → 0.29% del total): Se pueden eliminar las filas.

**PRODUCCION y PRECIO_CHACRA**(171 nulos → 3.5% del total): se puede eliminar las filas.

**COSECHA y RENDIMIENTO** (311-314 nulos → 6.4% del total): Podrían imputarse.

**Columnas con una cantidad considerable de nulos**:

**SIEMBRA** (894 nulos → 18.4% del total): Es alto, pero no extremo. Posible imputación basada en patrones.

**SUPERFICIE_PERDIDA** (4776 nulos → 98.1% del total): Prácticamente vacía. Podría eliminarse toda la columna si no es útil.

In [ ]:
#ELIMINADO COLUMNA SUPERFICIE PERDIDA, para no perder el 98.1%
df_limpiov1 = df_produccion.drop("SUPERFICIE_PERDIDA", axis=1)
print(df_limpiov1.isnull().sum()) #muestra la cantidad de valores nulos por columna

DEPARTAMENTO          0
PROVINCIA             0
DISTRITO              0
UBIGEO                0
PERIODO_AGRICOLA      0
CULTIVO               0
SUPERFICIE_VERDE     14
SIEMBRA             894
COSECHA             311
RENDIMIENTO         314
PRODUCCION          171
PRECIO_CHACRA       171
VALOR_PRODUCCION      0
FECHA_CORTE           0
dtype: int64


In [ ]:
#Eliminando valores nulos
df_limpiov2 = df_limpiov1.dropna()
print(df_limpiov2.isnull().sum())#deberia mostrar 0 nulos

DEPARTAMENTO        0
PROVINCIA           0
DISTRITO            0
UBIGEO              0
PERIODO_AGRICOLA    0
CULTIVO             0
SUPERFICIE_VERDE    0
SIEMBRA             0
COSECHA             0
RENDIMIENTO         0
PRODUCCION          0
PRECIO_CHACRA       0
VALOR_PRODUCCION    0
FECHA_CORTE         0
dtype: int64


In [ ]:
#información del dataset df limio version 2
print(df_limpiov2.info())

<class 'pandas.core.frame.DataFrame'>
Index: 3901 entries, 0 to 4868
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   DEPARTAMENTO      3901 non-null   object 
 1   PROVINCIA         3901 non-null   object 
 2   DISTRITO          3901 non-null   object 
 3   UBIGEO            3901 non-null   int64  
 4   PERIODO_AGRICOLA  3901 non-null   int64  
 5   CULTIVO           3901 non-null   object 
 6   SUPERFICIE_VERDE  3901 non-null   float64
 7   SIEMBRA           3901 non-null   float64
 8   COSECHA           3901 non-null   float64
 9   RENDIMIENTO       3901 non-null   float64
 10  PRODUCCION        3901 non-null   float64
 11  PRECIO_CHACRA     3901 non-null   float64
 12  VALOR_PRODUCCION  3901 non-null   float64
 13  FECHA_CORTE       3901 non-null   int64  
dtypes: float64(7), int64(3), object(4)
memory usage: 457.1+ KB
None


In [ ]:
print(df_limpiov2)

     DEPARTAMENTO PROVINCIA   DISTRITO  UBIGEO  PERIODO_AGRICOLA  \
0           CUSCO      ANTA  ANCAHUASI   80302              2020   
1           CUSCO      ANTA  ANCAHUASI   80302              2020   
2           CUSCO      ANTA  ANCAHUASI   80302              2020   
3           CUSCO      ANTA  ANCAHUASI   80302              2020   
4           CUSCO      ANTA  ANCAHUASI   80302              2020   
...           ...       ...        ...     ...               ...   
4861        CUSCO  URUBAMBA      YUCAY   81307              2018   
4864        CUSCO  URUBAMBA      YUCAY   81307              2018   
4866        CUSCO  URUBAMBA      YUCAY   81307              2018   
4867        CUSCO  URUBAMBA      YUCAY   81307              2018   
4868        CUSCO  URUBAMBA      YUCAY   81307              2018   

                                CULTIVO  SUPERFICIE_VERDE  SIEMBRA  COSECHA  \
0                               ALFALFA        190.470588    147.0    147.0   
1                     ARV

**NORMALIZACIÓN**

In [ ]:
# Ver el tipo de dato de cada columna
print(df_limpiov2.dtypes)

DEPARTAMENTO         object
PROVINCIA            object
DISTRITO             object
UBIGEO                int64
PERIODO_AGRICOLA      int64
CULTIVO              object
SUPERFICIE_VERDE    float64
SIEMBRA             float64
COSECHA             float64
RENDIMIENTO         float64
PRODUCCION          float64
PRECIO_CHACRA       float64
VALOR_PRODUCCION    float64
FECHA_CORTE           int64
dtype: object


In [ ]:
#Poner formato de fecha
# Transformar la columna de fecha a formato de fecha
df_limpiov2['FECHA_CORTE'] = pd.to_datetime(df_limpiov2['FECHA_CORTE'].astype(str), format="%Y%m%d")
print(df_limpiov2)

     DEPARTAMENTO PROVINCIA   DISTRITO  UBIGEO  PERIODO_AGRICOLA  \
0           CUSCO      ANTA  ANCAHUASI   80302              2020   
1           CUSCO      ANTA  ANCAHUASI   80302              2020   
2           CUSCO      ANTA  ANCAHUASI   80302              2020   
3           CUSCO      ANTA  ANCAHUASI   80302              2020   
4           CUSCO      ANTA  ANCAHUASI   80302              2020   
...           ...       ...        ...     ...               ...   
4861        CUSCO  URUBAMBA      YUCAY   81307              2018   
4864        CUSCO  URUBAMBA      YUCAY   81307              2018   
4866        CUSCO  URUBAMBA      YUCAY   81307              2018   
4867        CUSCO  URUBAMBA      YUCAY   81307              2018   
4868        CUSCO  URUBAMBA      YUCAY   81307              2018   

                                CULTIVO  SUPERFICIE_VERDE  SIEMBRA  COSECHA  \
0                               ALFALFA        190.470588    147.0    147.0   
1                     ARV

<ipython-input-33-2f952edd6786>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpiov2['FECHA_CORTE'] = pd.to_datetime(df_limpiov2['FECHA_CORTE'].astype(str), format="%Y%m%d")


In [ ]:
#EXPORTANDO ARCHIVO LIMPIO
df_limpiov2.to_csv("produccion_clean.csv", index=False, encoding="utf-8")